### PR TITLE
ref(receiver-constraints): Refactor and fixes.

### PR DESCRIPTION
**Pull request by hristoterezov**

### CODE REVIEW

File: react/features/large-video/actions.any.js [hunk 1]
```diff
@@ -72,13 +73,32 @@ export function selectParticipantInLargeVideo(participant: ?string) {
  *     resolution: number
  * }}
  */
-export function updateKnownLargeVideoResolution(resolution: number) {
+export function updateKnownLargeVideoResolution(resolution) {
```
Comment (jallamsetty1) [inline react/features/large-video/actions.any.js:76]: Why this change ? Wouldn't transition to TS be easier if we have types here ?

File: react/features/video-quality/subscriber.js [hunk 2]
```diff
@@ -289,110 +415,77 @@ function _updateReceiverVideoConstraints({ getState }) {
...
-            if (!visibleRemoteParticipants?.size && !largeVideoParticipantId) {
-                return;
-            }
+        if (getCurrentLayout(state) === LAYOUTS.STAGE_FILMSTRIP_VIEW && activeParticipantsSources.length > 0) {
+            const onStageSources = [ ...activeParticipantsSources ];
 
-            if (visibleRemoteParticipants?.size > 0) {
-                const qualityLevel = getVideoQualityForResizableFilmstripThumbnails(state);
-                const stageParticipantsLevel = getVideoQualityForStageThumbnails(state);
+            activeParticipantsSources.forEach(sourceName => {
```
Comment (jallamsetty1) [inline react/features/video-quality/subscriber.js:459]: Are `activeParticipantsSources` also included in the `visibleRemoteTrackSourceNames` ? If so, aren't we sending duplicate constraints here ?

File: react/features/video-quality/subscriber.js [hunk 3]
```diff
@@ -289,110 +415,77 @@ function _updateReceiverVideoConstraints({ getState }) {
...
-                visibleRemoteParticipants.forEach(participantId => {
-                    const isStageParticipant = activeParticipantsIds.find(id => id === participantId);
-                    const quality = Math.min(maxFrameHeight, isStageParticipant
-                        ? stageParticipantsLevel : qualityLevel);
+                receiverConstraints.constraints[sourceName] = { 'maxHeight': quality };
+            });
 
-                    receiverConstraints.constraints[participantId] = { 'maxHeight': quality };
-                });
+            if (screenshareFilmstripParticipantId) {
```
Comment (jallamsetty1) [inline react/features/video-quality/subscriber.js:468]: Same question as above for `screenshareFilmstripParticipantId` ? Isn't it already included in the visibleRemoteTrackSourceNames ?

File: react/features/video-quality/subscriber.js [hunk 4]
```diff
@@ -289,110 +415,77 @@ function _updateReceiverVideoConstraints({ getState }) {
...
+        if (getCurrentLayout(state) === LAYOUTS.STAGE_FILMSTRIP_VIEW && activeParticipantsSources.length > 0) {
+            const onStageSources = [ ...activeParticipantsSources ];
 
-            if (visibleRemoteParticipants?.size > 0) {
-                const qualityLevel = getVideoQualityForResizableFilmstripThumbnails(state);
-                const stageParticipantsLevel = getVideoQualityForStageThumbnails(state);
+            activeParticipantsSources.forEach(sourceName => {
+                const isScreenSharing = remoteScreenShares.includes(sourceName);
+                const quality
+                    = isScreenSharing && preferredVideoQuality >= MAX_VIDEO_QUALITY
```
Comment (jallamsetty1) [inline react/features/video-quality/subscriber.js:462]: What happens here when the preferredQuality is set to lower than 2160 and the send resolution on the SS stream being requested is higher than 2160 ? Does the bridge forward it anyways ?

**Review by jallamsetty1 — COMMENTED** [review]
This looks great! I have left some questions.

File: react/features/large-video/actions.any.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (hristoterezov) [reply react/features/large-video/actions.any.js:76]: We dropped flow and I'm not really sure when will the transition happen for this file. In addition to this we have the types in JS doc. 

I don't have strong opinion on the matter tho so if you insist I will revert it. Let me know!

File: react/features/video-layout/middleware.native.js [hunk 5]
```diff
@@ -18,6 +20,13 @@ MiddlewareRegistry.register(store => next => action => {
     case SET_CAR_MODE:
         dispatch(setVideoMuted(action.enabled, MEDIA_TYPE.VIDEO, VIDEO_MUTISM_AUTHORITY.CAR_MODE));
         break;
+    case CLIENT_RESIZED: {
+        const { clientHeight, clientWidth } = store.getState()['features/base/responsive-ui'];
+
+        // On mobile the large video should always fill the screen.
+        dispatch(setLargeVideoDimensions(clientHeight, clientWidth));
```
[the remark this answers is not in the document]
Comment (hristoterezov) [reply react/features/video-layout/middleware.native.js:27]: Yes this is used only for the receiver constraints. When I was discussing this with @saghul we agreed to take a shortcut here because the height of the large video will match the height of the screen in 99% of the cases. I think the only use case where it won't match is if we are displaying screen sharing and since we have pinch to zoom there better quality will be beneficial.

File: react/features/video-quality/constants.js [hunk 6]
```diff
@@ -19,6 +19,16 @@ export const VIDEO_QUALITY_LEVELS = {
     NONE: 0
 };
 
+/**
+ * Indicates unlimited video quality.
+ */
+export const VIDEO_QUALITY_UNLIMITED = 99999;
```
[the remark this answers is not in the document]
Comment (hristoterezov) [reply react/features/video-quality/constants.js:25]: Yes. Actually it turned out that the bridge was not handling -1 well. But I think this is fixed now and I should update it back to  -1.

The reason that this is not in the enum is because the enum is used in places dynamically and usually we are sorting it and using it to compare the height of the videos with the values. Basically if we have -1 there this whole approach needs to change and will introduce complications. In addition to this, since unlimited is kind of a special constraints, I think it is good idea to be separate.

File: react/features/video-quality/functions.js [hunk 7]
```diff
@@ -31,7 +8,7 @@ export function findNearestQualityLevel(videoQuality: number) {
  * ascending order.
  * @returns {number} The matching value from {@code VIDEO_QUALITY_LEVELS}.
  */
-export function getReceiverVideoQualityLevel(availableHeight: number, heightToLevel: Map<number, number>): number {
+export function getReceiverVideoQualityLevel(availableHeight, heightToLevel) {
```
[the remark this answers is not in the document]
Comment (hristoterezov) [reply react/features/video-quality/functions.js:11]: We dropped flow and I'm not really sure when will the transition happen for this file. In addition to this we have the types in JS doc.

I don't have strong opinion on the matter tho so if you insist I will revert it. Let me know!

File: react/features/video-quality/subscriber.js [hunk 8]
```diff
@@ -153,48 +139,158 @@ StateListenerRegistry.register(
...
+                maxVideoQualityChanged = true;
+                dispatch(
+                    setMaxReceiverVideoQualityForScreenSharingFilmstrip(
+                        newMaxRecvVideoQualityForScreenSharingFilmstrip));
+            }
         }
 
-        if (maxReceiverVideoQuality !== newMaxRecvVideoQuality) {
-            dispatch(setMaxReceiverVideoQuality(newMaxRecvVideoQuality));
+        if (!maxVideoQualityChanged && Boolean(displayTileView) !== Boolean(previousState.displayTileView)) {
```
[the remark this answers is not in the document]
Comment (hristoterezov) [reply react/features/video-quality/subscriber.js:290]: There's another subscriber that listens for changes in any of the max video quality variables and triggers the update. This will accommodate only the use case where only tile view state has changed and nothing else changes.

File: react/features/video-quality/subscriber.js (hunk 2 above)
Comment (hristoterezov) [reply react/features/video-quality/subscriber.js:459]: I think not always. IIRC some of they may be there and some not. But anyway even if there is overlap we will just override the value and not duplicate it which is exactly what we want because the activeParticipantSources are the once displayed on stage.

File: react/features/video-quality/subscriber.js (hunk 3 above)
Comment (hristoterezov) [reply react/features/video-quality/subscriber.js:468]: Same answer: it could be but I don't think this is an issue.

File: react/features/video-quality/subscriber.js (hunk 4 above)
Comment (hristoterezov) [reply react/features/video-quality/subscriber.js:462]: @bgrozev can explain this better.

But what I understood from him is that on VP9 the 3 layers are with the same resolution anyway and the constraints we set are mostly used to figure out what the user has selected as preferredVideoQuality... He was also saying it should be fine for VP8 too.

File: react/features/large-video/actions.any.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply react/features/large-video/actions.any.js:76]: I think it makes sense to leave it in to make the transition easier.

File: react/features/video-layout/middleware.native.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply react/features/video-layout/middleware.native.js:27]: I think the reverse is going to happen from what I understand about viewport sizes and screen resolutions. For example, the viewport size for iPad Pro is 1024 x 1366 while the actual resolution is about 2048 x 2732 Pixels since it has a pixel density of 2. So we will end up requesting lower resolutions. I don't think it matters much since we set the preferred video quality to 360p on mobile IIRC but again @saghul should be able to confirm.

File: react/features/video-quality/constants.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply react/features/video-quality/constants.js:25]: Ok makes sense about not having it as an enum.

File: react/features/video-quality/subscriber.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply react/features/video-quality/subscriber.js:290]: Aha, I missed that one in diff view. I see it now, thanks for the clarification.

File: react/features/video-quality/subscriber.js (hunk 4 above)
Comment (jallamsetty1) [reply react/features/video-quality/subscriber.js:462]: I would think all the layers would have same resolution for low fps sharing since the client sends only 1 stream/spatial layer irrespective of the codec. Only when high fps sharing is enabled, all the streams/layer with different resolutions will be enabled. Nonetheless, it will be worthwhile to test all the scenarios, especially when the SS sender does both low and high fps sharing and with both VP8 and VP9 codec.

File: react/features/video-layout/middleware.native.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (hristoterezov) [reply react/features/video-layout/middleware.native.js:27]: Now I see what you meant. 
So do you think in your example we need to request 2*height of video? 

Do we currently respect this anywhere? I don't think mobile web respects this currently as well.

You are correct that for react-native it doesn't matter in practice because preferred video quality is set to 360. But since this is an external parameter it is better to handle such things here.

Anyway if we don't handle this use case anywhere lets treat it as a new feature in a new PR? WDYT?

File: react/features/video-quality/subscriber.js (hunk 4 above)
Comment (hristoterezov) [reply react/features/video-quality/subscriber.js:462]: When I tested it was looking good for me! 

Anyway as I said you may check with @bgrozev what actually happens on the bridge side.

**Comment by hristoterezov** [comment]
@jallamsetty1 I think I addressed all the comments! I've also re-tested the PR and everything seems to work fine!

**Review by jallamsetty1 — APPROVED** [review]
Great work 🚀

**Review by saghul — APPROVED** [review]
